history_and_modals.md

```diff
@@ -1,158 +1,357 @@
+
+## What developers are doing today
+
+On desktop platforms, this problem is currently solved by listening for the `keydown` event, and closing the modal when the <kbd>Esc</kbd> key is pressed. Built-in platform APIs, such as [`<dialog>`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDialogElement/showModal), [fullscreen](https://developer.mozilla.org/en-US/docs/Web/API/Fullscreen_API), or [`<input type="date">`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date), will do this automatically. Note that this can be easy to get wrong, e.g. by accidentally listening for `keyup` or `keypress`.
+
+Additionally, note that the inability to prevent closing of `<dialog>`s via <kbd>Esc</kbd> makes them hard to use for some web developers, and requires them to roll their own custom dialog modals. _TODO: evidence of this would be good._
+
+On mobile platforms, getting the right behavior is significantly harder. First, platform-specific code is required, to do nothing on iOS and to capture the back button on Android. Next, capturing the back button on Android requires manipulating the history stack using the [history API](https://developer.mozilla.org/en-US/docs/Web/API/History_API). This is a poor fit for several reasons:
+
+- This UI state is non-navigational, i.e., the URL doesn't and shouldn't change when opening a modal. When the page is reloaded or shared, web developers generally don't want to automatically re-open the same modal.
+
+- It's very hard to determine when the modal state is popped out of history, because both moving forward and backward through history emit the same `popstate` event.
+
+- The `history.state` API, used to store the modal state, is rather fragile: user-initiated fragment navigation, as well as other calls to `history.pushState()` and `history.replaceState()`, can override it.
+
+- History navigation is not cancelable. Implementing "Are you sure you want to close this dialog without saving?" requires letting the history navigation complete, then potentially re-establishing the modal state if the user declines to close the modal.
+
+- A shared component that attempts to use the history API to implement these techniques can easily corrupt a web application's router.
+
+## Proposal
+
+The proposal is to introduce a new API, the `ModalCloseWatcher` class, which has the following basic API:
+
+```js
+// Constructor only works from within transient user activation;
+// otherwise, it throws a "NotAllowedError" DOMException.
+//
+// Only one such construction can be done per transient user activation,
+// but constructing does not consume user activation.
+const watcher = new ModalCloseWatcher();
```

Comment:
What do you think of passing the root of the modal UI to ModalCloseWatcher, that would better control when a "close" signal is being sent to that particular UI?

This would mirror IntersectionObserver, e.g.

Reply:
What would the implementation do with that information?

Reply:
This could also work as a partial abuse check. E.g. the element should be visible for back button to be intercepted.